### PR TITLE
Update waifu command

### PR DESCRIPTION
**Pull request by erksmit**
Fixes waifu command by using http2, and implements various improvements such as caching and pagination on search

### CODE REVIEW

File: bot/cogs/ext/commands/anime.py [hunk 1]
```diff
@@ -1,21 +1,21 @@
 """Father, forgive me for these sins."""
 
 import asyncio
+from typing import List, Union
```
Comment (JustAnyones) [inline bot/cogs/ext/commands/anime.py:4]: From X import Y imports should be placed on the second block

1st block is for general imports
2nd block is for from X import Y
3rd block is for custom Pidroid imports

File: bot/cogs/ext/commands/anime.py [hunk 2]
```diff
@@ -163,38 +183,34 @@ async def cancer_owo(self, ctx: Context, *, text: str = None):
...
+                if waifu_id:
+                    waifus.append(await api.fetch_waifu_by_id(waifu_id))
+                else:
+                    search_data = await api.search(selection)
+                    for search in search_data:
+                        if isinstance(search, WaifuSearchResult):
+                            waifus.append(search)
+                    if len(waifus) == 0:
+                        await ctx.reply(embed=error("Search did not find any waifus!"))
+                        return
```
Comment (JustAnyones) [inline bot/cogs/ext/commands/anime.py:207]: Can be simplified to just ```return await ctx.reply...```

File: bot/cogs/models/waifuListApi.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        """Initializes API instance."""
+        self._xsrf_token = None
+        self._csrf_token = None
+        self._forever_alone_session = None
+        self.client = httpx.AsyncClient(http2=True)
+        self.cache = {}
+
+    async def _acquire_tokens_for_forgery(self) -> None:
+        """Sends a GET request to dash page to acquire tokens for forgery."""
+        r = await self.client.get(BASE_URL + "/dash") # Because fuck you, that's why
```
Comment (JustAnyones) [inline bot/cogs/models/waifuListApi.py:193]: Profanity, please remove it

File: bot/cogs/models/waifuListApi.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
+from __future__ import annotations
```
Comment (JustAnyones) [inline bot/cogs/models/waifuListApi.py:1]: I would add a blank line after ```from __future__ import annotations``` import.

File: bot/cogs/models/waifuListApi.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    async def forged_headers(self) -> dict:
+        """Forged headers for making API calls."""
+        return {
+            "x-requested-with": "XMLHttpRequest",
+            "Referer": "https://mywaifulist.moe/dash",
+            "x-csrf-token": await self.csrf_token,
+            "x-xsrf-token": urllib.parse.unquote(await self.xsrf_token)
+        }
+
+    async def get_headers(self, json_as_string: bool = False) -> dict:
```
Comment (JustAnyones) [inline bot/cogs/models/waifuListApi.py:241]: Not used anywhere, this method is useless.

File: bot/cogs/models/waifuListApi.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        waifu_id = tree.xpath("//waifu-core")[0].attrib[':waifu-id']
+        waifu = await self.fetch_waifu_by_id(waifu_id)
+        self.cache[waifu_id] = waifu
+        return waifu
+
+    async def fetch_waifu_by_id(self, id: int) -> Waifu:
+        """Returns a waifu by the specified ID."""
+        if self.cache.get(id):
+            return self.cache.get(id)
+        else:
```
Comment (JustAnyones) [inline bot/cogs/models/waifuListApi.py:278]: Else can be commited since condition being true returns.

**Review by JustAnyones — CHANGES_REQUESTED** [review]
Resolve the conversations.

Also, I've noticed the new ``httpx`` dependency. Please specify it in the requirements file.

File: bot/cogs/models/waifuListApi.py (hunk 6 above)
Comment (erksmit) [reply bot/cogs/models/waifuListApi.py:278]: i just think it looks neater but youre the boss here

**Comment by erksmit** [comment]
issues resolved in e17082fcbd2244d52a2af201ce1837032e1d3db8

File: bot/cogs/ext/commands/anime.py [hunk 7]
```diff
@@ -1,11 +1,12 @@
 """Father, forgive me for these sins."""
 
 import asyncio
-from typing import List, Union
 import discord
 import random
 import re
 
+from typing import List, Union
+
```
Comment (JustAnyones) [inline bot/cogs/ext/commands/anime.py:9]: You added an extra line

**Review by JustAnyones — CHANGES_REQUESTED** [review]

File: bot/cogs/ext/commands/anime.py (hunk 7 above)
Comment (erksmit) [reply bot/cogs/ext/commands/anime.py:9]: import re

from typing import List, Union

from discord.ext import commands
isnt that what you wanted

Comment (JustAnyones) [reply bot/cogs/ext/commands/anime.py:9]: Let me spoon feed you

```py
import asyncio
import discord
import random
import re

from discord.ext import commands
from discord.ext.commands.context import Context
from discord.ext.commands.errors import BadArgument
from typing import List, Union

from constants import JESSE_ID, THEOTOWN_GUILD
from cogs.utils import http
from cogs.utils.embeds import create_embed, build_embed, error
from cogs.utils.paginators import ListPageSource, PidroidPages
from cogs.utils.parsers import truncate_string
from cogs.models.waifuListApi import MyWaifuListAPI, Waifu, WaifuSearchResult
from cogs.models.categories import RandomCategory
```

**Review by JustAnyones — CHANGES_REQUESTED** [review]
Also, lowercase the waifuListApi filename

File: bot/cogs/ext/commands/anime.py (hunk 7 above)
Comment (erksmit) [reply bot/cogs/ext/commands/anime.py:9]: now that just means it was incorrect before i even started

Comment (JustAnyones) [reply bot/cogs/ext/commands/anime.py:9]: Probable, anyway, it's good exercise to fix it right now :)

**Comment by erksmit** [comment]
do you mean removing the camelcase from the filename?

**Comment by JustAnyones** [comment]
Correct.

File: bot/cogs/ext/commands/anime.py [hunk 8]
```diff
@@ -47,6 +47,26 @@
...
+        self.embed = create_embed()
+
+    async def format_page(self, menu: PidroidPages, waifu: Union[Waifu, WaifuSearchResult]):
+        self.embed.clear_fields()
+        if isinstance(waifu, WaifuSearchResult):
+            waifu = await waifu.fetch_waifu()
+        self.embed.title = waifu.name
+        if waifu.is_nsfw and not menu.ctx.channel.is_nsfw():
+            self.embed.set_image(url="")
+            self.embed.description = 'this waifu is nsfw'
```
Comment (JustAnyones) [inline bot/cogs/ext/commands/anime.py:62]: Also, consider changing the following text to something like
```suggestion
            self.embed.description = 'This waifu is tagged as NSFW. In order to view the waifu, please use the command in an age-restricted channel.'
```

Comment (erksmit) [reply bot/cogs/ext/commands/anime.py:62]: considered it

**Comment by erksmit** [comment]
should be good now

File: bot/cogs/models/waifuListApi.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
-from __future__ import annotations
+from __future__ import annotation
```
Comment (JustAnyones) [inline bot/cogs/models/waifuListApi.py]: You dropped an s

Comment (erksmit) [reply bot/cogs/models/waifuListApi.py]: too slow

**Comment by erksmit** [comment]
all done

**Review by JustAnyones — APPROVED** [review]
Looks good to me

**Pull request merged**